modules/apps/client-extension/client-extension-web/src/main/java/com/liferay/client/extension/web/internal/type/deployer/CETDeployerImpl.java

```diff
@@ -56,15 +61,34 @@ public class CETDeployerImpl implements CETDeployer {
 
 	@Override
 	public List<ServiceRegistration<?>> deploy(CET cet) {
-		if (!Objects.equals(
+		if (Objects.equals(
 				cet.getType(),
-				ClientExtensionEntryConstants.TYPE_CUSTOM_ELEMENT) &&
-			!Objects.equals(
+				ClientExtensionEntryConstants.TYPE_CUSTOM_ELEMENT) ||
+			Objects.equals(
 				cet.getType(), ClientExtensionEntryConstants.TYPE_IFRAME)) {
 
-			return Collections.emptyList();
+			return _deployPortlet(cet);
 		}
 
+		if (Objects.equals(
+				cet.getType(), ClientExtensionEntryConstants.TYPE_JS_PACKAGE)) {
+
+			return _deployJSPackage(cet);
+		}
+
+		return Collections.emptyList();
+	}
+
+	@Activate
+	protected void activate(BundleContext bundleContext) {
+		_bundleContext = bundleContext;
+	}
+
+	private List<ServiceRegistration<?>> _deployJSPackage(CET cet) {
```

Comment:
By adding this in the `deploy()` operation, we assume the import map will be sent to all pages, all sites regardless whether it's actually needed. I think we don't have any sensible choice here at this point, because we probably don't want to manually tell where the import map for this module should be outputted and where not.

I looked for the `undeploy()` and seems we cover it generically :+1:

Reply:
We should definitely scope it by company, but for that we need to change the `JSImportmapsContributor` API.

Regarding showing only what is needed, in the page, that can't be done because the importmap is rendered once and forever right before anything else happens (so, when the user enters the page for the first time) and remains immutable forever.

It makes sense not to cover any other scenario since the importmap should be something very generic and small enough to not impact the performance. In other words: users shouldn't have hundreds of JS packages client extensions unless they want to resurrect the old issues we had with the AMD loader, when the registry was sent from the server to the browser and it weighted more than 2MB for some customers :-/ .